File: packages/benchmarks/rolodex/handlers/perfect.ts
import type { Handler, Conversation, Extraction, Resolution, GroundTruthWorld } from '../types';

export const perfectHandler: Handler = {
  name: 'Perfect (Oracle)',
  extract(conv: Conversation): Extraction {
    return {
      conversationId: conv.id,
      identities: conv.expected.identities.map(i => ({ ...i })),
      relationships: conv.expected.relationships.map(r => ({ ...r })),
      trustSignals: conv.expected.trustSignals.map(t => ({ ...t })),
      traces: ['Oracle: returned ground truth'],
      wallTimeMs: 0,
    };
  },
  resolve(_ext: Extraction[], world: GroundTruthWorld): Resolution {
    return {
      links: world.links.map(l => ({ entityA: l.entityA, entityB: l.entityB, confidence: 1.0, signals: [`Truth: ${l.reason}`] })),
      traces: ['Oracle: returned ground truth links'],
      wallTimeMs: 0,
    };
  },
};

File: packages/benchmarks/rolodex/python_bench/handlers/perfect.py
"""Perfect (Oracle) handler — returns ground truth for validation."""

from __future__ import annotations

from ..types import (
    Conversation,
    Extraction,
    GroundTruthWorld,
    IdentityExtraction,
    RelationshipExtraction,
    Resolution,
    ResolutionLink,
    TrustSignalExtraction,
)


class PerfectHandler:
    """Oracle handler that returns the exact ground truth.

    Used to validate the scoring harness: every metric should be 100 %.
    """

    name: str = "Perfect (Oracle)"

    async def setup(self) -> None:
        pass

    async def teardown(self) -> None:
        pass

    async def extract(
        self, conv: Conversation, world: GroundTruthWorld
    ) -> Extraction:
        return Extraction(
            conversation_id=conv.id,
            identities=[
                IdentityExtraction(
                    entity_id=i.entity_id,
                    platform=i.platform,
                    handle=i.handle,
                )
                for i in conv.expected.identities
            ],
            relationships=[
                RelationshipExtraction(
                    entity_a=r.entity_a,
                    entity_b=r.entity_b,
                    type=r.type,
                    sentiment=r.sentiment,
                )
                for r in conv.expected.relationships
            ],
            trust_signals=[
                TrustSignalExtraction(
                    entity_id=t.entity_id,
                    signal=t.signal,
                )
                for t in conv.expected.trust_signals
            ],
            traces=["Oracle: returned ground truth"],
            wall_time_ms=0.0,
        )

    async def resolve(
        self,
        extractions: list[Extraction],
        world: GroundTruthWorld,
    ) -> Resolution:
        return Resolution(
            links=[
                ResolutionLink(
                    entity_a=link.entity_a,
                    entity_b=link.entity_b,
                    confidence=1.0,
                    signals=[f"Truth: {link.reason}"],
                )
                for link in world.links
            ],
            traces=["Oracle: returned ground truth links"],
            wall_time_ms=0.0,
        )


perfect_handler = PerfectHandler()
